bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java

```diff
@@ -374,6 +306,91 @@ public void handlerRemoved(Thing thing, ThingHandler thingHandler) {
     }
 
     @Override
+    public void changeThingType(final Thing thing, final ThingTypeUID thingTypeUID,
+            final Configuration configuration) {
+        scheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                ThingUID thingUID = thing.getUID();
+                waitForRunningHandlerRegistrations(thingUID);
+                ThingType thingType = thingTypeRegistry.getThingType(thingTypeUID);
+                if (thingType == null) {
+                    logger.error("No thing type {} registered, cannot change thing type for thing {}",
+                            thingTypeUID.getAsString(), thingUID.getAsString());
+                    return;
+                }
+    
+                // Remove the ThingHandler, if any
+                final ThingHandlerFactory oldThingHandlerFactory = findThingHandlerFactory(thing.getThingTypeUID());
+                if (oldThingHandlerFactory != null) {
+                    unregisterHandler(thing, oldThingHandlerFactory);
+                    waitUntilHandlerUnregistered(thing, 60 * 1000);
+                } else {
+                    logger.debug("No ThingHandlerFactory available that can handle {}", thing.getThingTypeUID());
+                }
+    
+                // Set the new channels
+                List<Channel> channels = ThingFactoryHelper.createChannels(thingType, thingUID,
+                        configDescriptionRegistry);
+                ((ThingImpl) thing).setChannels(channels);
+    
+                // Set the given configuration
+                ThingFactoryHelper.applyDefaultConfiguration(configuration, thingType, configDescriptionRegistry);
+                ((ThingImpl) thing).setConfiguration(configuration);
+    
+                // Change the ThingType
+                ((ThingImpl) thing).setThingTypeUID(thingTypeUID);
+    
+                // Register the new Handler - ThingManager.updateThing() is going to take care of that
+                thingRegistry.update(thing);
+    
+                logger.debug("Changed ThingType of Thing {} to {}. New ThingHandler is {}.",
+                        thing.getUID().toString(), thing.getThingTypeUID(), thing.getHandler().toString());
```

Comment:
Second thought: yes, it is, the new handler is registered just the line before. No need to change.